### ISSUE TITLE
Relative `LISP` pathname fix breaks `LISP=sbcl`

### ISSUE THREAD

**Issue by mister-walter** [acl2/acl2#1254]
Commit 363cb9038a0fa24a432b67d62caed0b466df0947 breaks (at least on my system) the initial ACL2 image build process. I have confirmed that the previous commit (7d47351832c508699d3c9d17868b177ed5b89e86) builds properly.

The command I use to build ACL2 is `make LISP=sbcl`. Running this on commits that include 363cb9038a0fa24a432b67d62caed0b466df0947 results in my being dropped into the SBCL debugger. A relevant error message printed out is:
```
ACL2 Error in ACL2-INTERFACE:  Unable to obtain the truename of file
"sbcl".
```
I can provide additional build output if that would be helpful!

**Comment by MattKaufmann** [comment]
Yes, I'll need more detail.  The command "make LISP=sbcl" works for me on both Linux and Mac.  A start would be to send me your sbcl script and also make.log from the ACL2 build attempt.

**Comment by mister-walter** [comment]
I build SBCL from source with the following command:
```
 sh make.sh --without-immobile-space --without-immobile-code --without-compact-instance-header --fancy --dynamic-space-size=4Gb
```
I then use `sh install.sh` to install SBCL to the default location on my system, `/usr/local/bin/sbcl`.

[Here's my make.log.](https://github.com/acl2/acl2/files/6407432/make.log)

This issue is also reproducible on my Docker build: https://github.com/mister-walter/acl2-docker. The Dockerfile in that repo contains all of the commands that are needed to build SBCL in the same way I build it locally, as well as ACL2 (there is some additional complexity there with respect to making sure that the Docker image records which ACL2 commit it is building).

**Comment by MattKaufmann** [comment]
Thanks; make.log helped, as I think I've found the problem.  A first step is to confirm that I'm right about that; then I'll try to figure out what to do about it.  Here's a log on my Mac from executing (car sb-ext:*posix-argv*) in sbcl (it's similar on Linux):

    ~/acl2/acl2$ sbcl
    This is SBCL 2.1.2, an implementation of ANSI Common Lisp.
    More information about SBCL is available at <http://www.sbcl.org/>.

    SBCL is free software, provided as is, with absolutely no warranty.
    It is mostly in the public domain; some portions are provided under
    BSD-style licenses.  See the CREDITS and COPYING files in the
    distribution for more information.
    * (car sb-ext:*posix-argv*)
    "/Users/kaufmann/lisps/sbcl/current/src/runtime/sbcl"
    * 

May I see what you get?

**Comment by mister-walter** [comment]
Sure!
```
> sbcl                                                                                                                                           Fri 30 Apr 2021 03:09:26 PM EDT
This is SBCL 2.1.4, an implementation of ANSI Common Lisp.
More information about SBCL is available at <http://www.sbcl.org/>.

SBCL is free software, provided as is, with absolutely no warranty.
It is mostly in the public domain; some portions are provided under
BSD-style licenses.  See the CREDITS and COPYING files in the
distribution for more information.
* (car sb-ext:*posix-argv*)
"sbcl"
* 
```

**Comment by mister-walter** [comment]
Maybe `sb-ext:*runtime-pathname*` would be useful here? Looks like it's an undocumented SBCL feature that returns a path to the SBCL binary for the current process.

**Comment by MattKaufmann** [comment]
Excellent suggestion; thanks!  I hadn't known about that undocumented
variable, but now that you mention it I see it via google.  I'll make
an appropriate modification and push to github after a fresh
"everything" regression succeeds.  In case you want a patch in the
meantime, here's the diff.

    eld:/projects/acl2/acl2-git-scratch% git diff
    diff --git a/acl2-init.lisp b/acl2-init.lisp
    index 8dab10739..c95855149 100644
    --- a/acl2-init.lisp
    +++ b/acl2-init.lisp
    @@ -1919,7 +1919,26 @@ THISSCRIPTDIR=\"$( cd \"$( dirname \"$absdir\" )\" && pwd -P )\"
		 (concatenate 'string "$THISSCRIPTDIR/" core-name ".core")))
	 (with-open-file ; write to nsaved_acl2
	   (str sysout-name :direction :output)
    -      (let* ((prog (our-truename (car sb-ext:*posix-argv*) t)))
    +      (let ((prog
    +
    +; This is the program that will be invoked when starting ACL2.  We prefer an
    +; absolute pathname, rather than just "sbcl" or a relative pathname, so that
    +; ACL2 is invoked successfully even by those for whom "sbcl" points elsewhere
    +; (because of a different PATH).  At one point, between Versions 8.3 and 8.4,
    +; we applied our-truename to (car sb-ext:*posix-argv*).  But Andrew Walter
    +; reported a problem on 4/30/2021 (GitHub Issue #1254) that was due to nil
    +; being returned by that our-truename call, because (car sb-ext:*posix-argv*)
    +; is "sbcl" in his environment.  He suggested using undocumented SBCL global
    +; sb-ext:*runtime-pathname*, and we took that suggestion.
    +
    +             (or (our-truename sb-ext:*runtime-pathname* t)
    +
    +; If somehow the value above is nil, then we go with the called program.  We
    +; avoid calling our-truename in that case, since if "make LISP=sbcl" invokes a
    +; different sbcl than ./sbcl because "." is not at the front of the PATH at
    +; build time, then (our-truename "sbcl" t) would return the wrong path.
    +
    +                 (car sb-ext:*posix-argv*))))
	     (write-exec-file
	      str
	      ("~a~a~%"
    diff --git a/books/system/doc/acl2-doc.lisp b/books/system/doc/acl2-doc.lisp
    index 4db9201ac..88ea7eff9 100644
    --- a/books/system/doc/acl2-doc.lisp
    +++ b/books/system/doc/acl2-doc.lisp
    @@ -88196,7 +88196,9 @@ it."
      <p>Fixed builds that use a relative pathname for the LISP environment
      variable (for host Lisps CCL, SBCL, Allegro CL, and CMUCL; GCL and LispWorks
      didn't seem to have this problem).  Thanks to Mihir Mehta for bringing this
    - issue to our attention.</p>
    + issue to our attention; and thanks to Andrew Walter for reporting a problem
    + with our first solution for SBCL and proposing an alternative, which we
    + adopted.</p>

      <p>Fixed the process for running ACL2 without building an executable image.
      Some initialization that was missing from that process is now included.  Also,
    eld:/projects/acl2/acl2-git-scratch%

**Comment by mister-walter** [comment]
Confirmed that this was fixed by 0823213ce2d0b66f065f92f164ed35108a149366. Thanks again!
